link/kprobe.go

```diff
@@ -0,0 +1,231 @@
+// number of programs.
+func Kretprobe(symbol string) (*traceEvent, error) {
+	return kprobe(symbol, true)
+}
+
+// kprobe returns a TraceEvent backed by a k(ret)probe
+// of the given kernel symbol.
+func kprobe(symbol string, ret bool) (*traceEvent, error) {
+	if symbol == "" {
+		return nil, errors.New("symbol name cannot be empty")
+	}
+	if !rgxTraceEvent.MatchString(symbol) {
+		return nil, fmt.Errorf("symbol must be alphanumeric (with _ and -): %s", symbol)
+	}
+
+	// Use kprobe PMU if the kernel has it available.
+	tp, err := pmuKprobe(symbol, ret)
+	if err == nil {
+		return tp, nil
+	}
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("creating perf_kprobe PMU: %w", err)
+	}
+
+	// Use tracefs if kprobe PMU is missing.
+	tp, err = tracefsKprobe(traceGroup, symbol, ret)
+	if err != nil {
+		return nil, fmt.Errorf("creating trace event '%s' in tracefs: %w", symbol, err)
+	}
+
+	return tp, nil
+}
+
+// pmuKprobe returns a TraceEvent to be created using PMU (sysfs).
+func pmuKprobe(symbol string, ret bool) (*traceEvent, error) {
+	et, err := getPMUEventType("kprobe")
+	if err != nil {
+		return nil, err
+	}
+	return &traceEvent{
+		pmuID:     et,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe,
+		finalizer: nil, // no resources are created
+	}, nil
+}
+
+// tracefsKprobe returns a TraceEvent created by writing an entry to
+// <tracefs>/kprobe_events. On Close(), the entry is removed.
+// Repeated calls with the same arguments will yield TraceEvents without
+// finalizers, making sure only one removal attempt against
+// <tracefs>/kprobe_events is made during program teardown.
+func tracefsKprobe(group, symbol string, ret bool) (*traceEvent, error) {
+	te := traceEvent{
+		group:     group,
+		name:      symbol,
+		ret:       ret,
+		progType:  ebpf.Kprobe, // kernel only allows kprobe programs to attach
+		finalizer: nil,
+	}
+
+	// Before attempting to create a trace event through tracefs,
+	// check if an event with the same name already exists.
+	tid, err := getTraceEventID(group, kprobeTraceEventName(symbol, ret))
+	if err != nil && !errors.Is(err, ErrNotSupported) {
+		return nil, fmt.Errorf("getting trace event id: %w", err)
+	}
+	if err == nil {
+		// Kprobe trace event already exists, return a TraceEvent to the caller.
+		// Since this kprobe event was created by a prior call to this function,
+		// or even by another tool or process, it should not be removed by Close().
+		// Leave the finalizer empty.
+		te.finalizer = nil
+		te.tracefsID = tid
+		return &te, nil
+	}
+
+	// Kprobe trace event doesn't exist yet, create it using tracefs.
+	if err := createTraceFSKprobeEvent(group, symbol, ret); err != nil {
+		return nil, fmt.Errorf("creating kprobe event on tracefs: %w", err)
+	}
+
+	// Get the newly-created trace event's id.
+	tid, err = getTraceEventID(group, kprobeTraceEventName(symbol, ret))
+	if err != nil {
```

Comment:
probably we should call `closeTraceFSKprobeEvent` and cleanup before returning an error

Reply:
Hmm, not being able to read a newly-created trace event would be a freak error IMO. If this read fails, there's probably something unexpected going on in the kernel and we shouldn't try to manipulate state beyond that point.

Reply:
makes sense 👍🏻